### ISSUE TITLE
Non-applicable overloads should not be requested in subclasses

### ISSUE THREAD

**Issue by vnmabus** [python/mypy#13591]
In the following case, Mypy complains about an invalid signature for the `foo` method in the implementation class, as it was missing an overload.

However, the missing overload is not applicable to that subclass and should not be requested.

```python
from __future__ import annotations

from typing import overload, Generic, TypeVar

T = TypeVar("T")

class Interface(Generic[T]):
    @overload
    def foo(self: Interface[None]) -> int:
        ...

    @overload
    def foo(self, a: T) -> int:
        ...

    def foo(self, a: T | None = None) -> str | int:
        return 1


class Implementation(Interface[int]):
    
    def foo(self, a: int) -> int:
        return 2
```

_Originally posted by @vnmabus in https://github.com/python/mypy/issues/13106#issuecomment-1234390090_

**Comment by erictraut** [comment]
This appears to have been fixed in mypy 1.2.0.

**Comment by AlexWaygood** [comment]
Yeah, I think I remember the PR that fixed it, as well. Thanks!
